src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -739,18 +736,108 @@ public void visitSwitchExpression(JCSwitchExpression tree) {
+                    default -> {
+                        if (label.isExpression()) {
+                            JCExpression expr = (JCExpression) label;
+                            if (expr.hasTag(IDENT) && ((JCIdent) expr).sym.isEnum())
+                                constants.add(((JCIdent) expr).sym);
+                        } else {
+                            throw new AssertionError(label.getTag());
+                        }
+                    }
                 }
             }
+            for (Entry<Symbol, List<JCDeconstructionPattern>> e : categorizedDeconstructionPatterns.entrySet()) {
+                if (coversDeconstructionStartingFromComponent(pos, targetType, e.getValue(), 0)) {
+                    constants.add(e.getKey());
+                }
+            }
+            return constants;
+        }
+
+        private boolean coversDeconstructionStartingFromComponent(DiagnosticPosition pos,
+                                                                  Type targetType,
+                                                                  List<JCDeconstructionPattern> patterns,
+                                                                  int component) {
+            List<? extends RecordComponent> components = patterns.head.record.getRecordComponents();
+
+            if (components.size() == component) {
+                return true;
+            }
+
+            Type parameterizedComponentType = types.memberType(targetType, components.get(component));
+            List<JCPattern> nestedComponentPatterns = patterns.map(d -> d.nested.get(component));
+            Set<Symbol> nestedCovered = coveredSymbols(pos, parameterizedComponentType,
+                                                       nestedComponentPatterns);
+            Map<Symbol, List<JCDeconstructionPattern>> componentType2Patterns = new HashMap<>();
+            Set<Symbol> covered = new HashSet<>();
+
+            for (JCDeconstructionPattern subTypeCandidate : patterns) {
+                JCPattern nestedPattern = subTypeCandidate.nested.get(component);
+                Symbol currentPatternType;
+                switch (nestedPattern.getTag()) {
+                    case BINDINGPATTERN, PARENTHESIZEDPATTERN -> {
+                        PatternPrimaryType primaryPatternType =
+                                TreeInfo.primaryPatternType(nestedPattern);
+                        currentPatternType = primaryPatternType.type().tsym;
+                    }
+                    case DECONSTRUCTIONPATTERN -> {
+                        currentPatternType = ((JCDeconstructionPattern) nestedPattern).record;
+                    }
+                    default -> {
+                        throw Assert.error("Unexpected tree kind: " + nestedPattern.getTag());
+                    }
+                }
+                for (Symbol currentType : nestedCovered) {
+                    if (types.isSubtype(types.erasure(currentType.type),
```

Comment:
Not 100% what this test does

Reply:
I think this is i) from the domination relation:

> A record pattern with type R and record component pattern list L dominates another record pattern with type S and record component pattern list M if (i) the erasure of S is a subtype of the erasure of R, and (ii) every pattern, if any, in L dominates the corresponding pattern in M.

Reply:
I think this is i) from the domination relation:

> A record pattern with type R and record component pattern list L dominates another record pattern with type S and record component pattern list M if (i) the erasure of S is a subtype of the erasure of R, and (ii) every pattern, if any, in L dominates the corresponding pattern in M.

Reply:
You are right. It is the ii) which iteratively checks the component pattern list L.

Reply:
I now believe that the check is needed to properly classify patterns based on the type of the i-th component. That said, not sure this should be a subtyping check, or a type equality

Reply:
Ah - makes sense of course - I "just" needed a more convoluted example :-)